internal/pkg/service/buffer/worker/conditions/checker.go

```diff
@@ -0,0 +1,336 @@
+		if oldSlice.SliceKey != sliceKey {
+			return "", errors.Errorf(`cannot close slice "%s": unexpected export opened slice "%s"`, sliceKey.String(), oldSlice.FileKey)
+		}
+
+		export.OpenedSlice = model.NewSlice(oldSlice.FileKey, v.clock.Now(), oldSlice.Mapping, oldSlice.Number+1, oldSlice.StorageResource)
+		if newSlice, err := v.store.SwapSlice(ctx, &oldSlice); err == nil {
+			export.OpenedSlice = newSlice
+		} else {
+			return "", errors.Errorf(`cannot close slice "%s": cannot swap old and new slice: %w`, sliceKey.String(), err)
+		}
+
+		return "slice switched to the closing state", nil
+	})
+	return err
+}
+
+func (v *Checker) start(ctx context.Context, wg *sync.WaitGroup) <-chan error {
+	work := func(ctx context.Context, assigner *distribution.Assigner) (initDone <-chan error) {
+		lock := &sync.RWMutex{}
+		importConditions := make(conditionsMap)
+		openedSlices := make(slicesMap)
+
+		// Watch import conditions
+		init1 := v.schema.Configs().Exports().
+			GetAllAndWatch(ctx, v.etcdClient, etcd.WithPrevKV()).
+			SetupConsumer(v.logger).
+			WithForEach(func(events []etcdop.WatchEventT[model.ExportBase], header *etcdop.Header, restart bool) {
+				lock.Lock()
+				defer lock.Unlock()
+				for _, event := range events {
+					export := event.Value
+					if !assigner.MustCheckIsOwner(export.ReceiverKey.String()) {
+						// Another worker node handles the resource.
+						continue
+					}
+
+					switch event.Type {
+					case etcdop.CreateEvent, etcdop.UpdateEvent:
+						importConditions[export.ExportKey] = export.ImportConditions
+					case etcdop.DeleteEvent:
+						delete(importConditions, export.ExportKey)
+					default:
+						panic(errors.Errorf(`unexpected event type "%v"`, event.Type))
+					}
+				}
+			}).
+			StartConsumer(wg)
+
+		// Watch opened slices
+		init2 := v.schema.Slices().Opened().
+			GetAllAndWatch(ctx, v.etcdClient, etcd.WithPrevKV()).
+			SetupConsumer(v.logger).
+			WithForEach(func(events []etcdop.WatchEventT[model.Slice], header *etcdop.Header, restart bool) {
+				lock.Lock()
+				defer lock.Unlock()
+				for _, event := range events {
+					slice := event.Value
+					if !assigner.MustCheckIsOwner(slice.ReceiverKey.String()) {
+						// Another worker node handles the resource.
+						continue
+					}
+
+					switch event.Type {
+					case etcdop.CreateEvent, etcdop.UpdateEvent:
+						openedSlices[slice.SliceKey] = sliceInfo{
+							expiration: getCredentialsExpiration(slice),
+						}
+					case etcdop.DeleteEvent:
+						delete(openedSlices, slice.SliceKey)
+					default:
+						panic(errors.Errorf(`unexpected event type "%v"`, event.Type))
+					}
+				}
+			}).
+			StartConsumer(wg)
+
+		// Check conditions periodically
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			ticker := v.clock.Ticker(CheckInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					v.check(ctx, lock, importConditions, openedSlices)
+				}
+			}
```

Comment:
Periodicky sa kontroluju podmienky, vsetko je lokalne nacachovane a syncovane cez Watch API, takze to bude dobre skalovat.